### PR TITLE
Add schedule command (basic functionality)

### PR DESCRIPTION
**Pull request by lzan98**
Hi guys! This PR is quite a huge one so I will assign everybody to review this.

I will roughly summarise the main points of the schedule command, how the logic flows, what is implemented and what is missing due to further assistance/discussion required.

An interview consist of two parameters/fields, a `Candidate` and a `LocalDateTime`.
The input by the user for an interview's date and time must follow the format `dd/MM/yyyy HH:mm` (can be edited in future), and the date must not be in the past (does not sound logical to schedule an interview in the past!)

InterviewEndDateTime is defined as InterviewDateTime + 30 minutes.

I modelled the structure of `Interview` to be similar to `Candidate`, as follows: 
Candidate -> UniqueCandidateList -> AddressBook
Interview -> UniqueInterviewList -> InterviewSchedule

Whats important to note (and up for discussion) is the logic whenever the user wants to add an interview into the UniqueInterviewList. Right now, I set it such that an interview to be added must pass the following two checks:

1. `hasCandidate` There must not be another interview in the UniqueInterviewList with the same candidate. Logic for doing so is to prevent the user from accidentally scheduling the same candidate twice, when the candidate has not yet been interviewed!
2. `hasConflictingInterview` If there are no interviews with the same candidate found in the UniqueInterviewList, the interview to be added must not have conflicting time slots with existing interviews. An interview to be added is defined to have a conflicting time slot when its start time is within the start time and end time (30minutes period) of any existing interview, or if its' end time falls into the start time and end time of any existing interview.

Features not implemented due to the need for follow up:
- Display Interview Schedule on separate GUI panel instead of current feedback box
- Edit an interview (due to a need for confirmation on the point above)
- Delete an interview (same as point above)

P.S. Will add in more test cases over the weekend or after Monday's 2102 midterms, I think I've overestimated the scope of this feature and spent way too long on this PR T_T

### CODE REVIEW

File: src/main/java/seedu/address/MainApp.java [hunk 1]
```diff
@@ -76,21 +82,30 @@ public void init() throws Exception {
...
         try {
             addressBookOptional = storage.readAddressBook();
+            interviewListOptional = storage.readInterviewSchedule();
             if (!addressBookOptional.isPresent()) {
                 logger.info("Data file not found. Will be starting with a sample AddressBook");
             }
             initialData = addressBookOptional.orElseGet(SampleDataUtil::getSampleAddressBook);
+            //Does not make sense for sample interview list. Change to empty list
+            if (!interviewListOptional.isPresent()) {
+                logger.info("Data file for interviews not found. Will be starting with a sample InterviewSchedule");
```
Comment (domlimm) [inline src/main/java/seedu/address/MainApp.java:96]: Agreed. Good catch!

File: src/main/java/seedu/address/logic/commands/ScheduleCommand.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            "Interview for another candidate clashes with the proposed time slot!";
+
+    private final Index targetIndex;
+    private final LocalDateTime interviewDateTime;
+
+    /**
+     * Creates a ScheduleCommand to schedule the candidate at specified index for an
+     * interview on {@code LocalDateTime}
+     */
+    public ScheduleCommand(Index targetIndex, LocalDateTime interviewDateTime) {
```
Comment (domlimm) [inline src/main/java/seedu/address/logic/commands/ScheduleCommand.java:52]: Nice job on the `ScheduleCommand`! Well thought out! Team will definitely need to test it to sieve out anything to add.

Only comment: you might be missing the docs for `targetIndex` and `interviewDateTime`.

File: src/main/java/seedu/address/logic/commands/ViewCommand.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    @Override
+    public CommandResult execute(Model model) {
+        requireNonNull(model);
+        String message = model.getInterviewSchedule().getInterviewList().size() > 0
+                ? MESSAGE_SUCCESS
+                + model.getInterviewSchedule().getInterviewList()
+                .stream()
+                .sorted(Comparator.comparing(Interview::getInterviewDateTime))
+                .map(Object::toString)
+                .collect(Collectors.joining("\n"))
```
Comment (domlimm) [inline src/main/java/seedu/address/logic/commands/ViewCommand.java:31]: I might be wrong but these lines should be indented 4 spaces further. However, `checkstyle` did not pick it up i.e. Codecov passed. Hmmm! 🤔

File: src/main/java/seedu/address/logic/parser/ScheduleCommandParser.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+     * @throws ParseException If the command is not in a valid format.
+     */
+    private String[] parseDelimitedCommand(String command, String delimiter)
+            throws ParseException {
+        final String[] validDelimiter = command.split(delimiter);
+        if (validDelimiter.length < 2) {
+            throw new ParseException(MESSAGE_INVALID_COMMAND_FORMAT);
+        }
+        return validDelimiter;
+    }
```
Comment (domlimm) [inline src/main/java/seedu/address/logic/parser/ScheduleCommandParser.java:48]: Nice helper function!

File: src/main/java/seedu/address/model/InterviewSchedule.java [hunk 5]
```diff
@@ -0,0 +1,103 @@
+package seedu.address.model;
```
Comment (domlimm) [inline src/main/java/seedu/address/model/InterviewSchedule.java:1]: Looks good! I believe this file is similar to `src/main/java/seedu/address/model/AddressBook.java`. Glad there's a reference! Well thought out leon!

File: src/main/java/seedu/address/model/ModelManager.java [hunk 6]
```diff
@@ -112,8 +130,49 @@ public void setPerson(Candidate target, Candidate editedCandidate) {
...
+        interviewSchedule.addInterview(interview);
+        //updateFilteredPersonList(PREDICATE_SHOW_ALL_PERSONS);
+    }
+
+    /*@Override
+    public void setInterview(Interview target, Interview editedInterview) {
+        requireAllNonNull(target, editedInterview);
+
+        interviewSchedule.setInterview(target, editedInterview);
+    }*/
```
Comment (domlimm) [inline src/main/java/seedu/address/model/ModelManager.java:173]: And this too.

File: src/main/java/seedu/address/model/ModelManager.java [hunk 7]
```diff
@@ -112,8 +130,49 @@ public void setPerson(Candidate target, Candidate editedCandidate) {
...
+    @Override
+    public boolean hasConflictingInterview(Interview interview) {
+        requireNonNull(interview);
+        return interviewSchedule.hasConflictingInterview(interview);
+    }
+
+    /*@Override
+    public void deleteInterview(Interview target) {
+        interviewSchedule.removeInterview(target);
+    }*/
```
Comment (domlimm) [inline src/main/java/seedu/address/model/ModelManager.java:160]: I believe this and the one below are commented out for further discussion? 👍🏻

File: src/main/java/seedu/address/model/UserPrefs.java [hunk 8]
```diff
@@ -68,19 +79,22 @@ public boolean equals(Object other) {
...
+                interviewListFilePath);
     }
 
     @Override
     public String toString() {
         StringBuilder sb = new StringBuilder();
         sb.append("Gui Settings : " + guiSettings);
-        sb.append("\nLocal data file location : " + addressBookFilePath);
+        sb.append("\nLocal AddressBook data file location : " + addressBookFilePath);
+        sb.append("\nLocal InterviewSchedule data file location : " + interviewListFilePath);
```
Comment (domlimm) [inline src/main/java/seedu/address/model/UserPrefs.java:97]: Great on following the format! 😍

File: src/main/java/seedu/address/model/interview/Interview.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
+package seedu.address.model.interview;
```
Comment (domlimm) [inline src/main/java/seedu/address/model/interview/Interview.java:1]: Nice job leon!

File: src/main/java/seedu/address/model/interview/exceptions/ConflictingInterviewException.java [hunk 10]
```diff
@@ -0,0 +1,6 @@
+package seedu.address.model.interview.exceptions;
+
+public class ConflictingInterviewException extends RuntimeException {
+    public ConflictingInterviewException() {
+        super("Operation would result in conflicting interviews at the same time slot"); }
+}
```
Comment (domlimm) [inline src/main/java/seedu/address/model/interview/exceptions/ConflictingInterviewException.java:6]: Really neat of you creating an exception class! 😍

File: src/main/java/seedu/address/model/interview/UniqueInterviewList.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.interview.exceptions.ConflictingInterviewException;
+import seedu.address.model.interview.exceptions.DuplicateCandidateException;
+import seedu.address.model.interview.exceptions.InterviewNotFoundException;
+
+public class UniqueInterviewList implements Iterable<Interview> {
+    private final ObservableList<Interview> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Interview> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
```
Comment (domlimm) [inline src/main/java/seedu/address/model/interview/UniqueInterviewList.java:18]: I believe this is the `UI` part to integrate in 🤔

File: src/main/java/seedu/address/model/interview/exceptions/DuplicateCandidateException.java [hunk 12]
```diff
@@ -0,0 +1,10 @@
+package seedu.address.model.interview.exceptions;
+
+public class DuplicateCandidateException extends RuntimeException {
+    /**
+     * Signals that the operation will result in Interviews with duplicate candidates.
+     */
+    public DuplicateCandidateException() {
+        super("Operation would result in duplicate candidates");
+    }
+}
```
Comment (domlimm) [inline src/main/java/seedu/address/model/interview/exceptions/DuplicateCandidateException.java:10]: Nice one again on making it neat with a custom exception class!

File: src/main/java/seedu/address/storage/JsonAdaptedInterview.java [hunk 13]
```diff
@@ -0,0 +1,165 @@
+package seedu.address.storage;
```
Comment (domlimm) [inline src/main/java/seedu/address/storage/JsonAdaptedInterview.java:1]: From the looks of this file, I think it's still a WIP?

Lots of complaints from codecov 💢 let me take a look at how `JsonAdaptedPerson` is tested.

Think we need to discuss about the JSON format for the `Interview`.

I believe we need only `studentId` and an array of the `schedule`. And we probably only need to test these 2 fields for this file.

To KIV.

Again, good job leon!

File: src/test/java/seedu/address/logic/commands/CommandTestUtil.java [hunk 14]
```diff
@@ -42,11 +44,20 @@
...
     public static final String VALID_INTERVIEW_STATUS = "Pending";
     public static final String VALID_AVAILABILITY_AMY = "1,2,3,4,5";
     public static final String VALID_AVAILABILITY_BOB = "1,2,6,7";
 
+    public static final LocalDateTime VALID_ALICE_INTERVIEW_DATE_TIME =
+            LocalDateTime.of(2023, Month.FEBRUARY, 23, 15, 00);
+    public static final LocalDateTime VALID_BENSON_INTERVIEW_DATE_TIME =
+            LocalDateTime.of(2023, Month.DECEMBER, 01, 15, 00);
+    public static final LocalDateTime VALID_CARL_INTERVIEW_DATE_TIME =
+            VALID_ALICE_INTERVIEW_DATE_TIME.plusMinutes(15);
```
Comment (domlimm) [inline src/test/java/seedu/address/logic/commands/CommandTestUtil.java:58]: Java API serving `scheduling` well! Ha! 😄

File: src/test/java/seedu/address/logic/commands/ScheduleCommandTest.java [hunk 15]
```diff
@@ -0,0 +1,142 @@
+package seedu.address.logic.commands;
```
Comment (domlimm) [inline src/test/java/seedu/address/logic/commands/ScheduleCommandTest.java:1]: Really good job!

File: src/test/java/seedu/address/logic/commands/ViewCommandTest.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    @BeforeEach
+    public void setUp() {
+        model = new ModelManager(getTypicalAddressBook(), getTypicalInterviewSchedule(), new UserPrefs());
+        expectedModel = new ModelManager(getTypicalAddressBook(), getTypicalInterviewSchedule(), new UserPrefs());
+    }
+
+    @Test
+    public void execute_view_success() {
+        String expectedMessage = MESSAGE_SUCCESS + model.getInterviewSchedule().getInterviewList()
+                .stream().sorted(Comparator.comparing(Interview::getInterviewDateTime))
```
Comment (domlimm) [inline src/test/java/seedu/address/logic/commands/ViewCommandTest.java:40]: Nice use of the method reference operator (`::`)!

**Review by domlimm — APPROVED** [review]
I think first of all, really really, a darn good job leon!

Secondly, this PR is too huge!!! Guys, I think for everyone's sanity let's keep it under 500 lines incl. insertion and deletion for future PRs in only non-test files 😛 We can always break it up into smaller portions i.e. Creating more issues. But I believe this will be the biggest PR for tP.

This PR is huge but approved. I checked all files, they're independent of our existing codebase.

We gotta do intensive testing for this both manual and adding more test cases as codecov dropped for certain files. I believe it's just the formula codecov uses to calculate the drop in coverage (%).

I could chip in and help write the tests too. Let's discuss.

As of now, nothing from me. Let's look forward to the next discussion to speak about this feature.

Looks good to me! Again, tremendous effort leon! 💪🏻 👍🏻

File: src/main/java/seedu/address/storage/JsonAdaptedInterview.java (hunk 13 above)
Comment (lzan98) [reply src/main/java/seedu/address/storage/JsonAdaptedInterview.java:1]: Yes indeed Dom - great catch! While working on this I was wondering if it is possible to just store the `studentId` but ensure it works in the `toModelType()` method as well -> meaning to create an `Candidate` object from just the `studentId` field. However my brain was at its limit already so I left it as it is 😅

File: src/main/java/seedu/address/logic/commands/ScheduleCommand.java [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+ * Schedules a candidate identified using it's displayed index from the address book for an interview
+ * on a specified time slot.
+ */
+public class ScheduleCommand extends Command {
+
+    public static final String COMMAND_WORD = "schedule";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Schedules the candidate identified by the index number for an interview on given date and time.\n"
+            + "Time and date given must not be in the past.\n"
```
Comment (LeeEnEn) [inline src/main/java/seedu/address/logic/commands/ScheduleCommand.java:25]: Just nitpicking here, but maybe the `Time and date` could be in the same sequence, `Date and time`, as the other string messages.

**Review by LeeEnEn — APPROVED** [review]
Leon, I think you can change your role to be lead developer 😆

I don't see any glaring issue in this massive PR, but as Dom has mentioned, we need to do intensive testing to flush out any bugs, if there are any. 

This is insane man, well done Leon!! 👍👍

File: src/main/java/seedu/address/logic/commands/ScheduleCommand.java (hunk 17 above)
Comment (lzan98) [reply src/main/java/seedu/address/logic/commands/ScheduleCommand.java:25]: Good catch En En! Will fix it in subsequent commits

**Review by tiewweijian — APPROVED** [review]
I think Dom and En En has commented the portions already, and I think it's really a good job -- maybe after merging this in, we can work on including more test cases as well. Props to you @lzan98 !

**Review by charmainehly — APPROVED** [review]

**Pull request merged**